src/celeritas/grid/PolyEvaluator.hh

```diff
@@ -94,8 +94,8 @@ class PolyEvaluator
 //---------------------------------------------------------------------------//
 // Deduction Guides
 //---------------------------------------------------------------------------//
-template<typename T, unsigned int N>
-PolyEvaluator(Array<T, N> const&) -> PolyEvaluator<T, N - 1>;
+template<typename ArrayT>
+PolyEvaluator(ArrayT const&) -> PolyEvaluator<typename ArrayT::value_type, ArrayT::size() - 1>;
```

Comment:
This feels wrong. I'm not sure why C++ doesn't prefer this template match to the others...

Reply:
OK this was really subtle. Disabling the variadic deduction guide, we find:
```
/Users/seth/.local/src/celeritas-temp/test/celeritas/grid/PolyEvaluator.test.cc:39:30: error: no viable constructor or deduction guide for deduction of template arguments of 'PolyEvaluator'
    const auto eval_linear = PolyEvaluator{Array<int, 2>{10, 1}};
                             ^
/Users/seth/.local/src/celeritas-temp/src/celeritas/grid/PolyEvaluator.hh:57:39: note: candidate template ignored: couldn't infer template argument 'T'
    explicit CELER_CONSTEXPR_FUNCTION PolyEvaluator(Ts... coeffs)
                                      ^
/Users/seth/.local/src/celeritas-temp/src/celeritas/grid/PolyEvaluator.hh:67:30: note: candidate template ignored: couldn't infer template argument 'N'
    CELER_CONSTEXPR_FUNCTION PolyEvaluator(ArrayT const& coeffs)
                             ^
/Users/seth/.local/src/celeritas-temp/src/celeritas/grid/PolyEvaluator.hh:44:7: note: candidate template ignored: could not match 'PolyEvaluator' against 'Array'
class PolyEvaluator
      ^
/Users/seth/.local/src/celeritas-temp/src/celeritas/grid/PolyEvaluator.hh:98:1: note: candidate template ignored: substitution failure [with T = int]: deduced non-type template argument does not have the same type as the corresponding template parameter ('unsigned long' vs 'unsigned int')
PolyEvaluator(Array<T, N> const&) -> PolyEvaluator<T, N-1>;
^
```
the problem is that `Array`'s template parameter is `celeritas::size_type`, whereas the `PolyEvaluator` explicitly uses `unsigned int`. When compiling without CUDA, `size_type` takes the native integer type (64 for the 64-bit machines) so it can't match `unsigned long` against `unsigned int`.

Reply:
On my machine it doesn't, and it builds fine with the old way. I'm compiling with gcc and the tests were failing with clang. My best guess was that the variadic template was still a valid match for `Array<T,N>` and successfully constructed `PolyEvaluator<Array<T,N>,1>`, and then the template instantation failed on `PolyEvaluator<Array<T,N>,1>::operator(1)`.

Reply:
Ninja'd 🥷